### ISSUE TITLE
Automatic base positioning

### ISSUE THREAD

**Issue by mayacakmak** [hcrlab/stretch_web_interface#11]
### Problem description

The Stretch cannot move in the direction that its arm is extended. Because of that the user needs to drive towards a target then rotate 90 degrees towards the target to do manipulation. When the user fails to accurately position the base before rotating 90 degrees, they need to re-stow the arm, rotate back, drive closer, rotate again.. which is frustrating. This issue was observed in our user study and was mentioned by Henry Evans as a source of inefficiency and frustration (leading him to wish the Stretch has an omnidirectional base).

### Possible ways to address

Some ways to reduce the inefficiency and burden on the user:
* Automate base positioning by adding two buttons for rotating the base by 90degrees CW and CCW (this issue)
* Automate arm stowing and arm preparation to manipulate
* Visualize the reach of the arm during drive mode

@KaviMD All three of these are nicely modular functionalities that you can get started on. I'll create to additional issues for the later two items and reference this issue for the problem description.

**Comment by mayacakmak** [comment]
With the navigation actions still being triggered through overlay regions (for the targeted version of 2.00) this is how I would recommend adding the 90 degree rotation actions to the interface.

![Screen Shot 2021-06-24 at 2 41 38 PM](https://user-images.githubusercontent.com/3083404/123336264-7a145f00-d4fa-11eb-85ab-77dfe77a1526.png)

Later if/when we switch to a Beam-like driving interface we can re-think.

**Comment by KaviMD** [comment]
I've added the buttons and front-end code for rotating 90deg in both directions.

![image](https://user-images.githubusercontent.com/16945761/123469018-5f4bf400-d5a7-11eb-957a-49dafc022390.png)

For the backend, it looks like we are currently only sending angular velocity goals to the robot. Using stretch's IMU, I can put together a PID loop that would achieve the 90 rotation, but that feels like a pretty clunky solution.

What branch should I be committing this too @mayacakmak? The camera following and sim code is all setup, so I think the best option would be to merge [gazebo_sim](https://github.com/hcrlab/stretch_web_interface/tree/gazebo_sim) into master or latest and then create for the three tasks in this issue off of that, but I'm not sure

**Comment by KaviMD** [comment]
I implemented a function that turns the robot to by a specific angle offset, using a proportional control loop that generates repeated goals until the robot achieves the desired rotation. I'm getting rotation info from `tf`, which should work in the sim and on the robot.
https://github.com/hcrlab/stretch_web_interface/blob/%2311-decrease-manipulation-burden/robot/ros_connect.js#L356 (the p value should be tuned on the actual robot)

The loop stops automatically if the robot is translated or rotated via the web interface.

**Comment by mayacakmak** [comment]
Cool! I'm excited to test this on the robot.

**Comment by mayacakmak** [comment]
@KaviMD I'm testing this merge now. Nothing's happening when I trigger the 90 degree rotation, I'm getting this error that might be relevant. Let me know if you have any thoughts on what might this mean/how to fix.

![Screenshot from 2021-06-28 16-12-14](https://user-images.githubusercontent.com/3083404/123714879-08a31c00-d82c-11eb-8ee0-6818d17b9317.png)

**Comment by KaviMD** [comment]
I think the issue is with what transform I was getting the robot orientation from. Gazebo publishes as the root transform `odom`, which is what I was using before. (I used `rosrun tf view_frames` to find the root transform)

Updating `odom` this line: https://github.com/hcrlab/stretch_web_interface/blob/master/robot/ros_connect.js#L119, with whatever the non-sim root transform is should fix the problem. My guess is that its `world`.

**Comment by mayacakmak** [comment]
Yes! I also did `rosun tf view_frames` and tried to look for something like `robot_base` but couldn't find it. World is typically a fixed frame on the ground where the robot starts, not sure it's equivalent to `odom`

**Comment by KaviMD** [comment]
I think the base of the robot is `base_link`, which is what I'm using as the [fixed frame](https://github.com/hcrlab/stretch_web_interface/blob/master/robot/ros_connect.js#L103) when I get all the transforms. In order to get absolute orientation, I think that `base_link` would need to be compared to something fixed that is off the robot, like `world`, I may also be misunderstanding something about how transforms work though.

**Comment by mayacakmak** [comment]
That is correct! But typically we use world as the fixed frame and that allows the TFs that we read to be relative to that. Anyway, I tried that and it didn't work :-) it says "world" passed to lookUpTransform does not exist.. Also looking at the TF tree I don't see world or something separate from the robot, so perhaps the estimated pose of the robot is just not included in the TF. Let's ask Charlie what to do here.

**Comment by mayacakmak** [comment]
@elliston-f Quick question for you, since you're a collaborator on the repo ;) Is there an equivalent of `odom` TF frame for the (real) Stretch robot? Things were working in simulation with `odom` but not on the robot. If not, how can we access the base pose estimated with odometry on the Stretch? Thanks!

**Comment by mayacakmak** [comment]
Hmm, just noticed this line in the launch file:
`<param name="/stretch_driver/broadcast_odom_tf" type="bool" value="false"/>`
...will try changing it to true and go back to using `odom`.

**Comment by mayacakmak** [comment]
Okay, that resolved the issue of not having the odom frame! The feature not working yet though (multiple issues, will describe soon).

**Comment by hello-ck** [comment]
Hi everyone!

I have a couple of suggestions. 

@KaviMD, it's great that you've been working on feedback control for the mobile base, since it could help the robot better achieve its goals on some floors. I recently wrote a relevant [forum post](https://forum.hello-robot.com/t/base-motion-replay/222/4?u=hello-ck) that I recommend you read. Among other things, my post points to example code that uses a scan matcher with Stretch's laser range finder to provide improved odometry while controlling the mobile base. 

That said, I recommend starting with a simple open-loop rotation command to the robot. It won't be perfect, but it will do pretty well on most surfaces, since Stretch's low-level control uses wheel odometry to achieve the commanded angle. Also, the human operator will be able to make corrections after the autonomous attempt. 

For the real robot, a mobile base rotation command will rotate the robot by the commanded angle in radians. So, if you want to rotate Stretch by 90 degrees using wheel odometry, you won't need to consider reference frames. The web interface provides an example. The robot's browser executes rotation commands using the [baseTurn function](
https://github.com/hello-robot/stretch_web_interface/blob/2a7e345bef8a19549f618576d4fab8c09b8c452e/robot/ros_connect.js#L240) in `ros_connect.js`, which you could use to send a rotation command for ±π/2 radians. 

`baseTurn` uses the [generatePoseGoal function](https://github.com/hello-robot/stretch_web_interface/blob/2a7e345bef8a19549f618576d4fab8c09b8c452e/robot/ros_connect.js#L124) in `ros_connect.js` to move Stretch via the [FollowJointTrajectory](http://docs.ros.org/en/api/control_msgs/html/action/FollowJointTrajectory.html) action server provided by [stretch_ros](https://github.com/hello-robot/stretch_ros). If you're ever curious about the details, you can look at  [joint_trajectory_server.py](https://github.com/hello-robot/stretch_ros/blob/master/stretch_core/nodes/joint_trajectory_server.py). 

Outside of JavaScript, [keyboard_teleop](https://github.com/hello-robot/stretch_ros/blob/master/stretch_core/nodes/keyboard_teleop) provides a Python example that uses the FollowJointTrajectory action server to command the mobile base. It first [defines a command based on the user's keyboard input](https://github.com/hello-robot/stretch_ros/blob/92d077558c8e6553294f498c8a1e8913f5a4864d/stretch_core/nodes/keyboard_teleop#L222) and then [sends the command](https://github.com/hello-robot/stretch_ros/blob/92d077558c8e6553294f498c8a1e8913f5a4864d/stretch_core/nodes/keyboard_teleop#L286)  via the FollowJointTrajectory client in the [HelloNode class](https://github.com/hello-robot/stretch_ros/blob/92d077558c8e6553294f498c8a1e8913f5a4864d/hello_helpers/src/hello_helpers/hello_misc.py#L68).

I hope this helps!

Best wishes, 
Charlie

**Comment by KaviMD** [comment]
Thank you! The forum post was very helpful in understanding the different ways to get rotation data.

For some reason I thought that generatePoseGoal was setting velocity targets rather than position targets. I can definitely implement sending a command for much ±π/2 radians like you suggested, that seems like a much cleaner solution.

**Comment by mayacakmak** [comment]
That's super useful @hello-ck, thank you!

@KaviMD I was part way debugging the feedback control with odom fixed in the master branch yesterday. I can get back to it and fix this and test now if you like, but let me know if you're already working on it (just be sure to merge with master first to avoid conflicts), I can look into other stuff in the meantime. thanks!

**Comment by mayacakmak** [comment]
Indeed `generatePoseGoal` already using odometry greatly simplifies this feature.

**Comment by KaviMD** [comment]
I'm working on implementing the two necessary functions right now, if its okay I'll just push them to master, this should be a pretty small fix.

**Comment by mayacakmak** [comment]
Sounds good, let me know when it's ready to test.

**Comment by KaviMD** [comment]
Sorry for the delay, I just modified commands.js to use baseTurn instead of the turnAngleOffset. It doesn't turn exactly 90 degrees, but that may be a simulation artifact or require some on-robot tuning.

I'm in the process of removing all the code attached to turnAngleOffset and will push a commit for that in a bit.

**Comment by hello-ck** [comment]
I'm glad things are going well and you're close to an initial implementation to test on the robot! Three issues worth considering in the future follow:

1. **Rotational Speed Hazard:** The default rotational speed for the mobile base might be too high for regular use. You might want to send a rotational command with a lower speed. 
2. **Extended Arm Hazard:** If the telescoping arm is extended, rotating the mobile base can be hazardous. One way to handle this would be to refuse to autonomously rotate the mobile base if the arm is extended. Another way to handle this might be to reduce the rotational speed of the mobile base. For example, the linear speed of the farthest point (i.e., wrist or fingertips) might be kept constant. 
3. **Autonomous Motion Hazard:** If the operator detects an issue, it would be good for them to be able to stop the autonomous motion. Maya and I discussed the possibility of a dead man's switch approach. For example, autonomous motion might require the operator to hold down a mouse button while the mouse cursor is over a web interface button for the motion to continue. If the operator releases the mouse button, moves the mouse cursor off of the web interface button, or activates some sort of stop command, then the robot would cease its autonomous motion. For the mobile base, I believe sending a rotation command for 0 radians will stop the mobile base from rotating. [stretch_driver](https://github.com/hello-robot/stretch_ros/blob/master/stretch_core/nodes/stretch_driver) also provides a [stop_the_robot service](https://github.com/hello-robot/stretch_ros/blob/92d077558c8e6553294f498c8a1e8913f5a4864d/stretch_core/nodes/stretch_driver#L602). It would also be good for the robot to stop if the operator's browser disconnects.

Best wishes, 
Charlie

**Comment by mayacakmak** [comment]
@KaviMD Let me know if we have something ready to test for tomorrow.

Super important to think about the hazards that Charlie pointed out. 
* For 1: The speed from the earlier tests looked okay
* For 2: Let's add a warning pop up if the arm is extended when a rotation is triggered, with the options (a) Stow arm then rotate; (b) rotate anyways (e.g. if there's nothing around the robot the user might want to take the risk), (c) cancel rotation.
* For 3: We should definitely have a way of cancelling the rotation, I think you were already anticipating this from what I saw in the code. Keep holding the button until rotated 90 degrees might be a bit tiring and awkward (e.g. when you release it's unclear if you press again whether it does another new 90degrees or continue the previous one). But perhaps we can make it so that any click anywhere will stop the rotation. (Non critical, ideas for later) Would be nice to let the user know about that in an intuitive way how to stop the rotation (e.g. the cursor turns into a stop sign while the robot is moving and back to normal once it stops). Also would be nice to have a counter to visualize the robot's progress towards 90degrees while it is rotating.

**Comment by KaviMD** [comment]
The turn fix should already be committed to master (these [two](https://github.com/hcrlab/stretch_web_interface/commit/65cd83d70e3dde9a3f1d05be3b7861fec916cccb) [commits](https://github.com/hcrlab/stretch_web_interface/commit/3f23a3a8dce5ebe23e977ee50e6df2357dcba5fb))

I can work on adding 2 and 3. I did have something in place to cancel the rotation. Previously only base rotations or translations would interrupt the movement, but that is easy to extend to any event. (I [temporarily removed](https://github.com/hcrlab/stretch_web_interface/commit/3f23a3a8dce5ebe23e977ee50e6df2357dcba5fb) the interrupt functionality because it caused some errors with the new movement system, but I will add it back once its working)

**Comment by mayacakmak** [comment]
I just tested this and it works, yay! 

- [ ] Like you observed in simulation, it seems to rotate a bit less than 90degrees but pretty consistently; e.g. it gets back to its original rotation after one cw+ccw turn. So, for now, we can try to fix this by commanding a larger value I guess.
- [ ] It does not feel very fast but it is a little jumpy, it feels like it's moving in steps rather than a smooth continuous motion, so there could be some tuning of step size, velocity, frequency to smooth that.
- [ ] Clicking somewhere else on the interface stops it and executes the other command instead. This is fine but would be better to just stop and not execute another random command that the user happens to click on when they're trying to stop the continuous rotation.

For 2, it would be useful to first address Issue #13 (automatic stowing), add to that some way of tracking if the arm has moved away from its stowed pose and then implement this warning. Let me know if/when you want to work on that--I have a few minor fixes on other things but can get to is shortly.

**Comment by hello-ck** [comment]
Excellent! The following forum post seems relevant to the accuracy of the rotation: 

https://forum.hello-robot.com/t/inaccurate-base-rotations

**Comment by KaviMD** [comment]
@nickswalker These two files contain most of the 90 degree rotation code
https://github.com/hcrlab/stretch_web_interface/blob/928573a67e58ff9af36f4142845ee18f7b849e43/shared/commands.js#L530-L539
https://github.com/hcrlab/stretch_web_interface/blob/928573a67e58ff9af36f4142845ee18f7b849e43/robot/ros_connect.js#L422-L435
